File: articles/calculus-for-systems-modeling/articles/convergence-tests-and-the-discipline-of-infinite-approximation/c/convergence_test_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double geometric_sum(double a, double r, int n){
  double total = 0.0;
  for(int i=0;i<n;i++){ total += a * pow(r, i); }
  return total;
}

double p_series_sum(double p, int n){
  double total = 0.0;
  for(int i=1;i<=n;i++){ total += 1.0 / pow((double)i, p); }
  return total;
}

int main(void){
  double geo = geometric_sum(10.0, 0.6, 25);
  double geo_ref = 10.0 / (1.0 - 0.6);
  double p125 = p_series_sum(1.25, 10000);
  double p075 = p_series_sum(0.75, 10000);

  printf("series_name,test_used,n_terms,partial_sum,last_term,test_result,estimated_error\n");
  printf("geometric_r_0.6,geometric-series test,25,%.12f,%.12f,converges by geometric-series test,%.12f\n",geo,10.0*pow(0.6,24),geo_ref-geo);
  printf("p_series_1.25,p-series test,10000,%.12f,%.12f,converges,\n",p125,1.0/pow(10000.0,1.25));
  printf("p_series_0.75,p-series test,10000,%.12f,%.12f,diverges,\n",p075,1.0/pow(10000.0,0.75));
  return EXIT_SUCCESS;
}

File: articles/calculus-for-systems-modeling/articles/convergence-tests-and-the-discipline-of-infinite-approximation/fortran/convergence_test_audit.f90
program convergence_test_audit
  implicit none
  real(8) :: geo, geo_ref, p125, p075
  integer :: i

  geo = 0.0d0
  do i=0,24
    geo = geo + 10.0d0 * (0.6d0 ** i)
  end do
  geo_ref = 10.0d0 / (1.0d0 - 0.6d0)

  p125 = 0.0d0
  p075 = 0.0d0
  do i=1,10000
    p125 = p125 + 1.0d0 / (real(i,8) ** 1.25d0)
    p075 = p075 + 1.0d0 / (real(i,8) ** 0.75d0)
  end do

  print '(A)', 'series_name test_used n_terms partial_sum last_term test_result estimated_error'
  print '(A,1X,A,1X,I6,1X,F12.6,1X,F12.6,1X,A,1X,F12.6)', 'geometric_r_0.6','geometric',25,geo,10.0d0*(0.6d0**24),'converges',geo_ref-geo
  print '(A,1X,A,1X,I6,1X,F12.6,1X,F12.6,1X,A)', 'p_series_1.25','p-series',10000,p125,1.0d0/(10000.0d0**1.25d0),'converges'
  print '(A,1X,A,1X,I6,1X,F12.6,1X,F12.6,1X,A)', 'p_series_0.75','p-series',10000,p075,1.0d0/(10000.0d0**0.75d0),'diverges'
end program
